Compute Pressure API OT (#5421)
GoogleChrome/web.dev: src/site/content/en/blog/compute-pressure/index.md

---
layout: post

> title: Observing Compute Pressure

Author: "Compute pressure" is used pretty consistently within the proposal for the API, so I think we should stick with that.

> title: Observing Compute Pressure

Reviewer: I've never felt that was a compelling argument. Spec authors are professional engineers, not professional writers.

> title: Observing Compute Pressure

Author: True, but the way forward would probably be to work with the authors of the specification draft and get them to rename it: https://oyiptong.github.io/compute-pressure/

Failing that, I think we need to go with the name of record here.

> title: Observing Compute Pressure

Reviewer: That's getting to the point. This instance is not a reference to the API. It's a reference to the phenomenon being tracked. It's as if I were writing about a new wind mill ("Wonder Wind Mill") and capitalized every instance of wind mill ("Wind Mill") whether it referred to the product or not.

> title: Observing Compute Pressure

Author: Yup, I'm on board with the lower casing. My comment was regarding your "compute pressure" vs. "computational pressure" question, with the former matching the language in the proposed spec.

> title: Observing Compute Pressure

Reviewer: Oh! Sorry.

> subhead: An experimental API for reacting to changes in CPU utilization.

Author: I can switch to "use."

authors:
  - jeffposnick
date: 2021-05-21
description: An experimental API for reacting to changes in CPU utilization.
hero: image/FNkVSAX8UDTTQWQkKftSgGe9clO2/H9cVTFMj4cYmoIfGYQqd.jpg
alt: A photograph of a pressure gauge.
tags:
  - blog
  - capabilities
origin_trial:

>   url: t/k

Author: OT URL to be added once available.

are able to detect when a device is under this type of CPU stress, and can
selectively disable features or reduce their workloads in response.
Increasingly, there are web applications that would benefit from the same
information.

[Compute Pressure](https://github.com/oyiptong/compute-pressure/) is a new,
experimental API intended to allow web applications to gain insight into the CPU
utilization of the browser's device. The API is designed as a callback mechanism
that receives periodic updates, allowing developers to observe changes in CPU
utilization over time. A web app that uses this API can make informed decisions
about enabling or disabling heavyweight features in response to these
observations. It's a feedback mechanism that unlocks new avenues for adaptive
behavior, which can in turn lead to smoother, cooler experiences across a range
of hardware devices.

## Use cases

The motivating use cases are helping video conferencing and video game web
applications adapt to CPU utilization. The initial experimental implementation
of the API aims to support the following potential decisions:

### Video conferencing

> - Dynamically adjust the number of live video feeds displayed on the screen at

Reviewer: The original statement is in the imperative mood, which is a command to the reader. The new verb makes them gerunds (noun phrases basically). Please change all the bullets accordingly.

> - Dynamically adjust the number of live video feeds displayed on the screen at

Author: Okay, I've updated the additional bullet points as well.

  once.
- Adjust the video resolution or frame rate.
- Skip non-essential video processing, such as cosmetic filters.
- Disable non-essential audio processing, such as noise suppression.
- Skew the quality-vs-speed and size-vs-speed knobs towards "speed" in video and
  audio encoding.

### Video games

- Use lower-quality video and audio assets as needed.
- Disable optional effects like reflections, glare, or particle animations that
  don't affect gameplay.
- Tweak quality-vs-speed knobs in the game's rendering engine, affecting areas

>   like shadows quality, texture filtering, and view distance.

Reviewer: ```suggestion
  like shadow quality, texture filtering, and view distance.
```

The secondary use case is measuring the CPU resource consumption of a feature.
This ultimately supports the main goal of avoiding driving user devices into

> high CPU utilization. The API aims to help developers make the following types

Reviewer: ```suggestion
high CPU utilization. The API helps developers make the following types
```

Reviewer: This is called an inflated verb. It's rarely if ever needed.

of decisions:

- Compare the CPU consumption of alternative implementations of the same

>   feature, to find out which is more efficient. The API supports measuring CPU

Reviewer: ```suggestion
  feature to find out which is more efficient. The API supports measuring CPU
```

  utilization in the field via

>   [A/B tests](https://en.wikipedia.org/wiki/A/B_testing), because a feature's

Reviewer: ```suggestion
  [A/B tests](https://en.wikipedia.org/wiki/A/B_testing) because a feature's
```

  CPU utilization depends on the hardware it's running on. Most developers
  cannot afford performance measurement labs covering all the devices owned by
  their users.

> - Estimate the impact of enabling a feature on CPU consumption. This cost

Reviewer: ```suggestion
- Estimate the impact on CPU consumption of enabling a feature. This cost
```

  estimate feeds into the decisions outlined in the primary use case.

> ## Core concepts

Reviewer: Currently, the TOC on web.dev doesn't show different levels. There's also an implied commitment in any article that core concepts are covered. Therefore I recommend removing this heading and promoting the next two.

Author: Okay.

### CPU utilization

The CPU utilization of a device is the average of the utilization of all the
device's [CPU cores](https://en.wikipedia.org/wiki/Multi-core_processor),
represented in the API as a number between `0.0` and `1.0`.

A CPU core's utilization is the fraction of time that the core has been
executing code belonging to a thread, as opposed to being in an idle state.

A CPU utilization close to `0.0` indicates that the processor has been almost

> entirely idle. A CPU utilization close to `1.0` is very likely to lead to a
> device that is generating excessive heat. Web applications can help avoid bad

Reviewer: ```suggestion
entirely idle. A CPU utilization close to `1.0` is very likely to
generate excessive heat. Web applications can help avoiding bad
```

Reviewer: >help avoiding

I've been failing to mark situations like this. "Avoiding" is the start of the object portion of the sentence. (_subject verb_ [_object_]). Consequently, it must be a noun. A verb functioning as a noun must be a gerund and therefore a present participle verb. I don't know if you followed that, but I'm providing in case you want to dig into the logic.

user experiences by reducing their compute demands when the CPU utilization is
high.

### CPU clock speed

Modern CPU cores support a set of clock speeds. The device's firmware or

> operating system can set the core clock speed, in order to trade off the

Reviewer: ```suggestion
operating system can set the core clock speed in order to trade off the
```

available CPU computational resources with power consumption.

From a user experience standpoint, the following are the most interesting clock
speeds:

- The minimum clock speed, which results in the lowest power consumption.
- The base clock speed, which results in a moderate level of power consumption
  and heat generation. This tends to be the speed advertised in marketing
  materials.
- The maximum clock speed, sometimes marketed as "turbo boost." Running at this
  speed causes unsustainable amounts of heating. It can only be used for short
  periods of time, to satisfy bursts in demand for computing power.

When a device's CPU utilization gets high, the device may increase clock speeds
across its CPU cores, in an attempt to meet the demand for computing power. As
the speeds exceed the base clock speed, the elevated power consumption increases
the CPU's temperature. At some point, the device enters a
[thermal throttling regime](https://en.wikipedia.org/wiki/Dynamic_frequency_scaling),

> where the CPU clock speed is reduced, in order to bring the temperature down.

Reviewer: ```suggestion
where the CPU clock speed is reduced in order to bring the temperature down.
```

  } else {
    // The CPU is running at normal or reduced speed.
  }

  if (update.cpuUtilization >= 0.9) {
    // CPU utilization is over 90%.
  } else if (update.cpuUtilization >= 0.75) {
    // CPU utilization is over 75%.
  } else if (update.cpuUtilization >= 0.5) {
    // CPU utilization is over 50%.
  } else {
    // CPU utilization is under 50%.
  }
}

const observer = new ComputePressureObserver(callback, {
  // Thresholds divide the interval [0.0 .. 1.0] into ranges.
  cpuUtilizationThresholds: [0.5, 0.75, 0.9],
  // The minimum clock speed is 0.0, and the maximum speed is 1.0.
  // 0.5 represents the base clock speed.
  cpuSpeedThresholds: [0.5],
});

// Begin observing changes and triggering callbacks:
observer.observe();

// ...later on, to stop observing changes:
observer.unobserve();
```

### Exposing limited information via thresholds and buckets

One of the key API design goals is to provide enough information for an
application to make useful decisions based on compute pressure, while not
exposing information that might
[harm a user's privacy](https://github.com/oyiptong/compute-pressure/blob/main/security-privacy-self-assessment.md).

> With this in mind, the API reports back normalized, unitless values between

Reviewer: ```suggestion
With this in mind, the API reports normalized, unitless values between
```

`0.0` and `1.0` instead of more identifiable values like clock speeds in GHz.
The values passed to the callback fall into one of the "buckets" defined by the
thresholds used to configure the observer. The actual value will be the midpoint

> between the upper and lower bounds for the bucket—in other words, they will be

Reviewer: ```suggestion
between the upper and lower bounds for the bucket. In other words, they will be
```

Author: One day you'll let me sneak an em dash through 😄

values determined by how you configure the observer, and not based on the actual
device.

To give an example, the configuration in the preceding section uses
`cpuUtilizationThresholds: [0.5, 0.75, 0.9]`. This configuration sets up a total
of four possible buckets that the reported values might fall into, covering the
ranges 0.0 to 0.5, 0.5 to 0.75, 0.75 to 0.9, and 0.9 to 1.0. To represent bucket

> membership, the value that is reported to the callback is the average between

Reviewer: ```suggestion
membership, the value reported to the callback is the average between
```

  </figcaption>
</figure>

Here's how some "real" CPU utilization values, as reported by the operating
system to the browser, would be represented as values passed to your callback:

- "Real" 0% utilization would map to the 0.0 to 0.5 bucket, and would be
  reported as `cpuUtilization: 0.25`, since 0.25 is the average of 0.0 and 0.5.
- "Real" 40% utilization would similarly map to the 0.0 to 0.5 bucket, and be
  reported as `cpuUtilization: 0.25`
- "Real" 60% utilization would map to the 0.5 to 0.75 bucket, and be reported as
  `cpuUtilization: 0.625`
- "Real" 80% utilization would map to the 0.75 to 0.9 bucket, and be reported as
  `cpuUtilization: 0.825`

The same threshold and bucket system applies to the exposed CPU speed values.

{% Aside %}
The values reported by the API reflect the overall state of the CPU,
due to utilization across all applications currently running. Therefore, even if
your own web application is almost entirely idle, you may see high utilization
and clock speed values reported, due to the activities of other applications.
{% endAside %}

## Try a demo

When accessed on a browser that supports the origin trial (currently Chrome 92

> or later), this [demo page](https://oyiptong.github.io/compute-pressure-demo/)

Author: This URL will likely be updated once the OT starts and we can move to a new repo.

> or later), this [demo page](https://oyiptong.github.io/compute-pressure-demo/)

Reviewer: ```suggestion
or later), [this demo](https://oyiptong.github.io/compute-pressure-demo/)
```

will display an ongoing readout of your device's CPU utilization and speed.

## API status

<div class="w-table-wrapper">

| Step                                     | Status                   |
| ---------------------------------------- | ------------------------ |
| 1. Create explainer                      | [Complete](https://github.com/oyiptong/compute-pressure/blob/main/README.md)    |
| 2. Create initial draft of specification | [In Progress](https://oyiptong.github.io/compute-pressure/)      |
| 3. Gather feedback & iterate on design   | [In progress](#) |
| 4. **Origin trial**                      | [In progress](#ot)    |
| 5. Launch                                | Not started              |

</div>

## API Availability

### Enabling support during the origin trial phase {: #ot }

> Starting in Chrome 92, the Compute Pressure API can be used as part of an origin

Author: M92 needs to be confirmed.
